File: backend/src/mentor/api/v1/curriculum.py
"""Curriculum endpoints."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from mentor.api.deps import SessionDep
from mentor.application.curriculum import CurriculumService
from mentor.application.curriculum.service import LessonWithProgress
from mentor.domain.curriculum.progress import LessonStatus
from mentor.domain.errors import ValidationError
from mentor.infrastructure.repositories import LessonProgressRepository

router = APIRouter(prefix="/curriculum", tags=["curriculum"])


class LessonSummary(BaseModel):
    slug: str
    title: str
    summary: str
    est_minutes: int
    order_in_module: int
    key_concepts: list[str]
    status: LessonStatus


class ModuleSummary(BaseModel):
    id: str
    order: int
    title: str
    summary: str
    est_minutes: int
    completed_count: int
    total_count: int
    is_complete: bool
    lessons: list[LessonSummary]


class FigureDTO(BaseModel):
    key: str
    caption: str


class QuizQuestionDTO(BaseModel):
    prompt: str
    options: list[str]
    correct_index: int
    explanation: str


class LessonResponse(BaseModel):
    slug: str
    module_id: str
    order_in_module: int
    title: str
    summary: str
    body_md: str
    est_minutes: int
    key_concepts: list[str]
    figures: list[FigureDTO]
    quiz: list[QuizQuestionDTO]
    status: LessonStatus


class MarkRequest(BaseModel):
    status: LessonStatus


def _service(session: SessionDep) -> CurriculumService:
    return CurriculumService(LessonProgressRepository(session))


@router.get("/overview", response_model=list[ModuleSummary])
async def overview(session: SessionDep) -> list[ModuleSummary]:
    modules = await _service(session).overview()
    return [
        ModuleSummary(
            id=m.module.id,
            order=m.module.order,
            title=m.module.title,
            summary=m.module.summary,
            est_minutes=m.module.est_minutes,
            completed_count=m.completed_count,
            total_count=m.total_count,
            is_complete=m.is_complete,
            lessons=[
                LessonSummary(
                    slug=lp.lesson.slug,
                    title=lp.lesson.title,
                    summary=lp.lesson.summary,
                    est_minutes=lp.lesson.est_minutes,
                    order_in_module=lp.lesson.order_in_module,
                    key_concepts=list(lp.lesson.key_concepts),
                    status=lp.progress.status if lp.progress else LessonStatus.NOT_STARTED,
                )
                for lp in m.lessons
            ],
        )
        for m in modules
    ]


@router.get("/lessons/{slug:path}", response_model=LessonResponse)
async def get_lesson(slug: str, session: SessionDep) -> LessonResponse:
    try:
        lp = await _service(session).get(slug)
    except ValidationError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return _lesson_response(lp)


def _lesson_response(lp: LessonWithProgress) -> LessonResponse:
    return LessonResponse(
        slug=lp.lesson.slug,
        module_id=lp.lesson.module_id,
        order_in_module=lp.lesson.order_in_module,
        title=lp.lesson.title,
        summary=lp.lesson.summary,
        body_md=lp.lesson.body_md,
        est_minutes=lp.lesson.est_minutes,
        key_concepts=list(lp.lesson.key_concepts),
        figures=[FigureDTO(key=f.key, caption=f.caption) for f in lp.lesson.figures],
        quiz=[
            QuizQuestionDTO(
                prompt=q.prompt,
                options=list(q.options),
                correct_index=q.correct_index,
                explanation=q.explanation,
            )
            for q in lp.lesson.quiz
        ],
        status=lp.progress.status if lp.progress else LessonStatus.NOT_STARTED,
    )


@router.post("/lessons/{slug:path}/progress", response_model=LessonResponse)
async def mark_progress(slug: str, body: MarkRequest, session: SessionDep) -> LessonResponse:
    service = _service(session)
    try:
        await service.mark(slug, body.status)
        lp = await service.get(slug)
    except ValidationError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return _lesson_response(lp)

File: frontend/src/api/curriculum.ts
import { z } from 'zod';

import { apiGet, apiPost } from './client';

const status = z.enum(['not_started', 'in_progress', 'completed']);
export type LessonStatus = z.infer<typeof status>;

const lessonSummary = z.object({
  slug: z.string(),
  title: z.string(),
  summary: z.string(),
  est_minutes: z.number(),
  order_in_module: z.number(),
  key_concepts: z.array(z.string()),
  status,
});
export type LessonSummary = z.infer<typeof lessonSummary>;

const moduleSummary = z.object({
  id: z.string(),
  order: z.number(),
  title: z.string(),
  summary: z.string(),
  est_minutes: z.number(),
  completed_count: z.number(),
  total_count: z.number(),
  is_complete: z.boolean(),
  lessons: z.array(lessonSummary),
});
export type ModuleSummary = z.infer<typeof moduleSummary>;

const lessonResponse = z.object({
  slug: z.string(),
  module_id: z.string(),
  order_in_module: z.number(),
  title: z.string(),
  summary: z.string(),
  body_md: z.string(),
  est_minutes: z.number(),
  key_concepts: z.array(z.string()),
  figures: z.array(z.object({ key: z.string(), caption: z.string() })),
  quiz: z.array(
    z.object({
      prompt: z.string(),
      options: z.array(z.string()),
      correct_index: z.number(),
      explanation: z.string(),
    })
  ),
  status,
});
export type LessonResponse = z.infer<typeof lessonResponse>;
export type LessonFigureData = LessonResponse['figures'][number];
export type QuizQuestion = LessonResponse['quiz'][number];

export async function getOverview(): Promise<ModuleSummary[]> {
  return apiGet('/curriculum/overview', z.array(moduleSummary));
}

export async function getLesson(slug: string): Promise<LessonResponse> {
  return apiGet(`/curriculum/lessons/${slug}`, lessonResponse);
}

export async function markLesson(
  slug: string,
  newStatus: LessonStatus
): Promise<LessonResponse> {
  return apiPost(
    `/curriculum/lessons/${slug}/progress`,
    { status: newStatus },
    lessonResponse
  );
}
